macrosynergy/pnl/transaction_costs.py

```diff
@@ -0,0 +1,255 @@
+        for txn_ticker in [tcost_n, rcost_n, size_n, tcost_l, rcost_l, size_l]
+    ]
+    found_tickers = list(set(df["cid"] + "_" + df["xcat"]))
+    if not set(expected_tickers).issubset(set(found_tickers)):
+        raise ValueError(
+            "The dataframe is missing the following tickers: "
+            + ", ".join(set(expected_tickers) - set(found_tickers))
+        )
+
+    return None
+
+
+def get_diff_index(df_wide: pd.DataFrame, freq: str = "D") -> pd.Index:
+    df_diff = df_wide.diff(axis=0)
+    change_index = df_diff.index[((df_diff.abs() > 0) | df_diff.isnull()).any(axis=1)]
+    return change_index
+
+
+def extrapolate_cost(
+    trade_size: Number,
+    median_size: Number,
+    median_cost: Number,
+    pct90_size: Number,
+    pct90_cost: Number,
+) -> Number:
+    err_msg = "{k} must be a number > 0"
+    for k, v in [
+        ("trade_size", trade_size),
+        ("median_size", median_size),
+        ("median_cost", median_cost),
+        ("pct90_size", pct90_size),
+        ("pct90_cost", pct90_cost),
+    ]:
+        if not isinstance(v, Number):
+            raise TypeError(err_msg.format(k=k))
+        if v <= 0:
+            raise ValueError(err_msg.format(k=k))
+
+    if trade_size <= median_size:
+        cost = median_cost
+    else:
+        b = (pct90_cost - median_cost) / (pct90_size - median_size)
+        cost = median_cost + b * (trade_size - median_size)
+    return cost
+
+
+class SparseCosts(object):
+    """
+    Interface to query transaction statistics dataframe.
+    """
+
+    def __init__(
+        self,
+        df: QuantamentalDataFrame,
+    ) -> None:
+        if not isinstance(df, QuantamentalDataFrame):
+            raise TypeError("df must be a QuantamentalDataFrame")
+        df_wide = qdf_to_ticker_df(df)
+        self._all_fids = get_fids(df)
+        change_index = get_diff_index(df_wide)  # drop rows with no change
+        df_wide = df_wide.loc[change_index]
+        self.df_wide = df_wide
+
+    def get_costs(self, fid: str, real_date: str) -> pd.DataFrame:
+        assert fid in self._all_fids
+        cost_names = [col for col in self.df_wide.columns if col.startswith(fid)]
+        if not cost_names:
+            raise ValueError(f"Could not find any costs for {fid}")
+        df_loc = self.df_wide.loc[:real_date, cost_names]
+        last_valid_index = df_loc.last_valid_index()
+        return df_loc.loc[last_valid_index]
+
+
+class TransactionCosts(object):
+    """
+    Interface to query transaction statistics dataframe.
+    """
+
+    DEFAULT_ARGS = dict(
+        tcost_n="BIDOFFER_MEDIAN",
+        rcost_n="ROLLCOST_MEDIAN",
+        size_n="SIZE_MEDIAN",
+        tcost_l="BIDOFFER_90PCTL",
+        rcost_l="ROLLCOST_90PCTL",
+        size_l="SIZE_90PCTL",
+    )
+
+    @classmethod
+    def download(self) -> "TransactionCosts":
+        df = download_transaction_costs()
+        return TransactionCosts(df=df, fids=get_fids(df), **self.DEFAULT_ARGS)
```

Comment:
Shouldn't this be:
```python
@classmethod
def download(cls) -> "TransactionCosts":
    df = download_transaction_costs()
    return cls(df=df, fids=get_fids(df), **self.DEFAULT_ARGS)
```
to follow standard notion?

Reply:
Also, personally I think it is more readable if defined after the `__init__` method, so you first give the user the standard way of instantiating the class, and then secondly the alternative (`download`).

Reply:
I like it though - even as I prefer downloading the data separately.